### PR TITLE
Apply some improvements to cache handling

### PR DESCRIPTION
**Pull request by bburnichon**
refs #1414

### CODE REVIEW

File: lib/Alchemy/Phrasea/Core/Configuration/AccessRestriction.php [hunk 1]
```diff
@@ -60,6 +61,25 @@ public function isDataboxAvailable(\databox $databox)
     }
 
     /**
+     * @param \databox[] $databoxes
+     * @return \databox[]
+     */
+    public function filterDataboxAvailable(array $databoxes)
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Configuration/AccessRestriction.php]: `filterAvailableDataboxes` feels more englishy.

File: lib/Alchemy/Phrasea/Databox/DataboxRepository.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    /**
+     * @param int $id
+     * @return \databox|null
+     */
+    public function find($id)
+    {
+        try {
+            $databox = $this->factory->create($this->app, (int)$id);
+        } catch (NotFoundHttpException $exception) {
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Databox/DataboxRepository.php]: You should explain with a small comment why there is this weird exception catching here.

File: lib/Alchemy/Phrasea/Databox/DataboxRepository.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+     * @return \databox[]
+     */
+    public function findAll()
+    {
+        try {
+            $rows = $this->appbox->get_data_from_cache(\appbox::CACHE_LIST_BASES);
+            if (!is_array($rows)) {
+                throw new \UnexpectedValueException('Expects rows to be an array');
+            }
+        } catch(\Exception $e) {
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Databox/DataboxRepository.php]: Meh

File: lib/Alchemy/Phrasea/Databox/CachedDataboxRepository.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+class CachedDataboxRepository implements DataboxRepositoryInterface
+{
+    /** @var DataboxRepositoryInterface */
+    private $repository;
+    /** @var \appbox */
+    private $appbox;
+    /** @var DataboxHydrator */
+    private $hydrator;
+
+    public function __construct(DataboxRepositoryInterface $repository, \appbox $appbox, DataboxHydrator $hydrator)
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Databox/CachedDataboxRepository.php]: No need for a full `appbox`, `$app[‘cache’]` is sufficient here.

File: lib/Alchemy/Phrasea/Databox/CachedDataboxRepository.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        $this->appbox = $appbox;
+        $this->hydrator = $hydrator;
+    }
+
+    public function find($id)
+    {
+        $rows = $this->fetchRows();
+
+        if (isset($rows[$id])) {
+            return $this->hydrator->hydrateRow($id, $rows[$id]);
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Databox/CachedDataboxRepository.php]: Why not call `Factory::create()`?

File: lib/Alchemy/Phrasea/Databox/CachedDataboxRepository.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+        return $this->repository->find($id);
+    }
+
+    public function findAll()
+    {
+        $rows = $this->fetchRows();
+
+        if (is_array($rows)) {
+            return $this->hydrator->hydrateRows($rows);
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Databox/CachedDataboxRepository.php]: Why not call `DataboxFactory::create()` in a foreach?

File: lib/Alchemy/Phrasea/Databox/CachedDataboxRepository.php [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * (c) 2005-2015 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace Alchemy\Phrasea\Databox;
+
+use Alchemy\Phrasea\Application;
+use Alchemy\Phrasea\Cache\Exception;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Databox/CachedDataboxRepository.php]: `as CacheException`?

File: lib/Alchemy/Phrasea/Databox/DataboxFactory.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * @param array $rows
+     * @throws NotFoundHttpException when Databox could not be retrieved from Persistence layer
+     * @return \databox[]
+     */
+    public function createMany(array $rows)
+    {
+        $databoxes = [];
+
+        foreach ($rows as $id => $raw) {
```
Comment (aztech-dev) [inline lib/Alchemy/Phrasea/Databox/DataboxFactory.php]: raw ?

File: lib/classes/databox.php [hunk 9]
```diff
@@ -1575,4 +1512,34 @@ public function isRegistrationEnabled()
...
+        $this->labels['de'] = $row['label_de'];
+        $this->labels['nl'] = $row['label_nl'];
+    }
+
+    /**
+     * Return an array that can be used to restore databox.
+     *
+     * @return array
+     */
+    public function getAsRow()
```
Comment (aztech-dev) [inline lib/classes/databox.php]: Why not toArray() ?

**Pull request merged**